{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0acec5f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "from typing import List, Dict, Optional\n",
    "import sys\n",
    "import os\n",
    "sys.path.append(os.path.abspath(os.path.join(os.getcwd(), '..')))\n",
    "import backend.knowledgespace_api as k"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "374a6038",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['scr_003105_neurondb_currents',\n",
       " 'scr_002721_gensat_geneexpression',\n",
       " 'scr_014194_icg_ionchannels',\n",
       " 'scr_006131_hba_atlas',\n",
       " 'scr_017571_dandi',\n",
       " 'scr_017041_sparc_old',\n",
       " 'scr_005069_brainminds',\n",
       " 'scr_017041_sparc',\n",
       " 'scr_003510_cil_images',\n",
       " 'scr_007271_modeldb_models',\n",
       " 'scr_013705_neuroml_models',\n",
       " 'scr_016433_conp',\n",
       " 'scr_002978_aba_expression',\n",
       " 'scr_002145_neuromorpho_modelimage',\n",
       " 'scr_017612_ebrains',\n",
       " 'scr_014306_bbp_cellmorphology',\n",
       " 'scr_005031_openneuro',\n",
       " 'scr_006274_neuroelectro_ephys']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "datasource_list = k.list_datasources()\n",
    "datasource_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "5da0e225",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "27"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = k.get_datasource_keys(\"scr_006274_neuroelectro_ephys\")\n",
    "keywords = a.get(\"keywords\", [])\n",
    "len(keywords)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "ae9948d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['scr_017571_dandi', 'scr_006274_neuroelectro_ephys', 'scr_002721_gensat_geneexpression', 'scr_017041_sparc', 'scr_002145_neuromorpho_modelimage', 'scr_005069_brainminds', 'scr_007271_modeldb_models', 'scr_014194_icg_ionchannels', 'scr_003510_cil_images', 'scr_016433_conp', 'scr_013705_neuroml_models', 'scr_017612_ebrains', 'scr_005031_openneuro', 'scr_017041_sparc_old', 'scr_014306_bbp_cellmorphology', 'scr_006131_hba_atlas', 'scr_003105_neurondb_currents', 'scr_002978_aba_expression']\n",
      "[22, 27, 34, 9, 39, 12, 28, 34, 43, 13, 40, 32, 170, 9, 20, 21, 22, 25]\n"
     ]
    }
   ],
   "source": [
    "key_len=[]\n",
    "for i in range(len(datasource_list)):\n",
    "    datasource = datasource_list[i]\n",
    "    key_len.append(len(k.get_datasource_keys(datasource).get(\"keywords\", [])))\n",
    "print(datasource_list)\n",
    "print(key_len)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "68fbac35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'total_count': 835,\n",
       " 'current_page': 0,\n",
       " 'results': [{'dc': {'title': 'Physiological Properties and Behavioral Correlates of Hippocampal Granule Cells and Mossy Cells',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000003/draft',\n",
       "    'description': 'Data from \"Physiological Properties and Behavioral Correlates of Hippocampal Granule Cells and Mossy Cells\" Senzai, Buzsaki, Neuron 2017. Electrophysiology recordings of hippocampus during theta maze exploration.'},\n",
       "   'id': '000003'},\n",
       "  {'dc': {'title': 'A NWB-based dataset and processing pipeline of human single-neuron activity during a declarative memory task',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000004/draft',\n",
       "    'description': 'A challenge for data sharing in systems neuroscience is the multitude of different data formats used. Neurodata Without Borders: Neurophysiology 2.0 (NWB:N) has emerged as a standardized data format for the storage of cellular-level data together with meta-data, stimulus information, and behavior. A key next step to facilitate NWB:N adoption is to provide easy to use processing pipelines to import/export data from/to NWB:N. Here, we present a NWB-formatted dataset of 1863 single neurons recorded from the medial temporal lobes of 59 human subjects undergoing intracranial monitoring while they performed a recognition memory task. We provide code to analyze and export/import stimuli, behavior, and electrophysiological recordings to/from NWB in both MATLAB and Python. The data files are NWB:N compliant, which affords interoperability between programming languages and operating systems. This combined data and code release is a case study for how to utilize NWB:N for human single-neuron recordings and enables easy re-use of this hard-to-obtain data for both teaching and research on the mechanisms of human memory.'},\n",
       "   'id': '000004'},\n",
       "  {'dc': {'title': 'Electrophysiology data from thalamic and cortical neurons during somatosensation',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000005/draft',\n",
       "    'description': 'intracellular and extracellular electrophysiology recordings performed on mouse barrel cortex and ventral posterolateral nucleus (vpm) in whisker-based object locating task.'},\n",
       "   'id': '000005'},\n",
       "  {'dc': {'title': 'Mouse anterior lateral motor cortex (ALM) in delay response task',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000006/draft',\n",
       "    'description': 'Extracellular electrophysiology recordings performed on mouse anterior lateral motor cortex (ALM) in delay response task. Neural activity from two neuron populations, pyramidal track upper and lower, were characterized, in relation to movement execution. Some files, as originally (re)distributed from e.g. http://datasets.datalad.org/?dir=/labs/svoboda/Economo_2018 were found to be broken and would not be able available among reorganized files under sub-* directories.'},\n",
       "   'id': '000006'},\n",
       "  {'dc': {'title': 'A cortico-cerebellar loop for motor planning',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000007/draft',\n",
       "    'description': 'Extracellular recording in ALM'},\n",
       "   'id': '000007'},\n",
       "  {'dc': {'title': 'Phenotypic variation within and across transcriptomic cell types in mouse motor cortex',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000008/draft',\n",
       "    'description': 'We used Patch-seq to combine patch-clamp recording, biocytin staining, and single-cell RNA sequencing of over 1300 neurons in adult mouse motor cortex, providing a comprehensive morpho-electric annotation of almost all transcriptomically defined neural cell types. Contained in this dandiset are the intracellular electrophysiological recordings. See Dandiset #35 for an additional dataset, recorded under the physiological temperature. '},\n",
       "   'id': '000008'},\n",
       "  {'dc': {'title': 'Maintenance of persistent activity in a frontal thalamocortical loop',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000009/draft',\n",
       "    'description': 'We recorded spikes from the ALM and thalamus during tactile discrimination  with a delayed directional response. Here we show that, similar to ALM  neurons, thalamic neurons exhibited selective persistent delay activity  that predicted movement direction. Unilateral photoinhibition of delay  activity in the ALM or thalamus produced contralesional neglect.  Photoinhibition of the thalamus caused a short-latency and near-complete  collapse of ALM activity. Similarly, photoinhibition of the ALM diminished  thalamic activity. Our results show that the thalamus is a circuit hub in  motor preparation and suggest that persistent activity requires reciprocal  excitation across multiple brain areas.'},\n",
       "   'id': '000009'},\n",
       "  {'dc': {'title': 'A motor cortex circuit for motor planning and movement',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000010/draft',\n",
       "    'description': 'Data from \"A motor cortex circuit for motor planning and movement\" Li et al. Nature 2015'},\n",
       "   'id': '000010'},\n",
       "  {'dc': {'title': 'Robust neuronal dynamics in premotor cortex during motor planning',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000011/draft',\n",
       "    'description': 'Data from \"Robust neuronal dynamics in premotor cortex during motor planning\" Nature 2016'},\n",
       "   'id': '000011'},\n",
       "  {'dc': {'title': 'Kriegstein2020',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000012/draft',\n",
       "    'description': 'Data from the Kriegstein Lab as part of the BICCN'},\n",
       "   'id': '000012'},\n",
       "  {'dc': {'title': 'Low-noise encoding of active touch by layer 4 in the somatosensory cortex',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000013/draft',\n",
       "    'description': 'Data from \"Low-noise encoding of active touch by layer 4 in the somatosensory cortex\" Hires, Gutnisky et al. Elife 2015'},\n",
       "   'id': '000013'},\n",
       "  {'dc': {'title': 'A Map of Anticipatory Activity in Mouse Motor Cortex',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000015/draft',\n",
       "    'description': 'Data from \"A Map of Anticipatory Activity in Mouse Motor Cortex\" Chen et al. Neuron 2017'},\n",
       "   'id': '000015'},\n",
       "  {'dc': {'title': 'Excitatory and inhibitory subnetworks are equally selective during decision-making and emerge simultaneously during learning',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000016/draft',\n",
       "    'description': 'This package contains data, in NWB (Neurodata Without Borders) format, from the 4 mice included in \"Najafi, Farzaneh, Gamaleldin F. Elsayed, Robin Cao, Eftychios Pnevmatikakis, Peter E. Latham, John P. Cunningham, and Anne K. Churchland. \"Excitatory and inhibitory subnetworks are equally selective during decision-making and emerge simultaneously during learning.\" Neuron 105, no. 1 (2020): 165-179.” Each NWB file represents the data and metadata associated with one recording session. In each NWB file, the metadata related to the session (mouse name, session date/time, lab/institution name, etc.) can be found under \"general\". Information related to ROI-segmentation such as ROI mask, ROI type (excitatory or inhibitory), poor or good quality, etc. can be found under \"modules/Image-Segmentation/pln-seg\". Trial information (e.g. start, end times, trial types, trial outcomes, etc.) can be found under \"trials\". Recorded trial-segmented neuronal responses aligned to different time event (e.g. stimulus start, animal choice, etc.) can be found under \"modules/ Trial-based-Segmentation\". A jupyter notebook presenting in detail how to work with NWB files is provided at https://github.com/ttngu207/najafi-2018-nwb/blob/master/notebooks/Najafi-2018_example.ipynb.'},\n",
       "   'id': '000016'},\n",
       "  {'dc': {'title': 'Distributed coding of choice, action and engagement across the mouse brain',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000017/draft',\n",
       "    'description': 'Data from \"Distributed coding of choice, action and engagement across the mouse brain\" Steinmetz et. al Nature 2019'},\n",
       "   'id': '000017'},\n",
       "  {'dc': {'title': 'Mouse Spinal Cord Ephys Dataset',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000018/draft',\n",
       "    'description': 'Electrophysiological data obtained in 1U01MH116990-01'},\n",
       "   'id': '000018'},\n",
       "  {'dc': {'title': 'Human ECoG speaking consonant-vowel syllables',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000019/draft',\n",
       "    'description': 'The enclosed data is collected using a high-density 256-channel electrocorticography (ECoG) array implanted in human patients during treatment for epilepsy. The subjects are reading aloud consonant-vowel syllables from a list. The data was collected by Dr. Edward Chang and Dr. Kristofer Bouchard at the University of California, San Francisco, and curated by Dr. Kristofer Bouchard and Dr. Benjamin Dichter.\\n\\n\\nData is organized by subject ID, and each file is a continuous recording session in Neurodata Without Borders (NWB) 2.0 format. Voltage traces are included for each of the recorded 256 channels. Microphone signal was recorded at the time but is removed for HIPAA compliance. Detailed hand-marked annotations are provided which mark what syllable was said, and the times of the start, consonant-vowel transition, and end of each syllable. A rest-period time is also included when the subject was silent, which can be used as a baseline. Hand-marked anatomical labels are included for electrodes in the relevant brain regions. All dates have been removed for HIPAA compliance and replaced with Jan 1, 1900.'},\n",
       "   'id': '000019'},\n",
       "  {'dc': {'title': 'Patch-seq recordings from mouse visual cortex',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000020/draft',\n",
       "    'description': 'Whole-cell Patch-seq recordings from neurons of the mouse visual cortex from the Allen Institute for Brain Science, released in June 2020. The majority of cells in this dataset are GABAergic interneurons, but there are also a small number of glutamatergic neurons from layer 2/3 of the mouse visual cortex.'},\n",
       "   'id': '000020'},\n",
       "  {'dc': {'title': '20191003_AIBS_mouse_ecephys_brain_observatory_1_1',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000021/draft',\n",
       "    'description': 'Allen Institute October 2019 Mouse extracellular electrophysiology data approximately matching two-photon brain observatory stimulus set. \\n\\nFor more information, see https://portal.brain-map.org/explore/circuits/visual-coding-neuropixels\\n\\nData are subject to Allen Institute terms of use, available at: http://www.alleninstitute.org/legal/terms-use/'},\n",
       "   'id': '000021'},\n",
       "  {'dc': {'title': '20191003_AIBS_mouse_ecephys_functional_connectivity',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000022/draft',\n",
       "    'description': 'Allen Institute October 2019 Mouse extracellular electrophysiology data, collected under analogous conditions to the two-photon Brain Observatory imaging experiments, with mice shown a subset of stimuli with a higher number of repeats. \\n\\nFor more information, see https://portal.brain-map.org/explore/circuits/visual-coding-neuropixels\\n\\nData are subject to Allen Institute Terms of Use policy, available at: http://www.alleninstitute.org/legal/terms-use/\\n'},\n",
       "   'id': '000022'},\n",
       "  {'dc': {'title': 'Patch-seq recordings from human cortex (June 2020)',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000023/draft',\n",
       "    'description': 'Whole-cell Patch-seq recordings from neurons of the human neocortex from the Allen Institute for Brain Science, released in June 2020.'},\n",
       "   'id': '000023'},\n",
       "  {'dc': {'title': 'LFP & MUA from BF',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000024/draft',\n",
       "    'description': 'billateral LFP recordings from barrel field of rats during bi- and unilateral whiskers stimulation'},\n",
       "   'id': '000024'},\n",
       "  {'dc': {'title': 'Example intracellular ephys data from LNMC & BBP',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000025/draft',\n",
       "    'description': 'This is a single session showing example intracellular electrophysiology data collected at the Laboratory of Neural Microcircuitry, EPFL. This session holds a variety of different stimulation patterns. \\n\\nIt also serves as an example of the hierarchy of metadata tables (/general/intracellular_ephys/) and groups (/data_organization) storing sweeps of a complex experimental protocol.\\n\\nLoad this file in pynwb with load_namespaces = True.\\n'},\n",
       "   'id': '000025'},\n",
       "  {'dc': {'title': 'Human brain cell census for BA 44/45',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000026/draft',\n",
       "    'description': 'Magnetic resonance imaging (MRI) is used to establish a macroscopic reference coordinate system of laminar and cytoarchitectural boundaries. Cell counting is obtained with both traditional immunohistochemistry, to provide a stereological gold standard, and with a custom-made inverted confocal light-sheet fluorescence microscope (LSM) for 3D imaging at cellular resolution. Finally, polarization-sensitive optical coherence tomography (PSOCT) enables registration of the distorted histological cell typing obtained with LSM to the MRI-based atlas coordinate system.'},\n",
       "   'id': '000026'},\n",
       "  {'dc': {'title': 'Test dataset for testing dandi-cli.',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000027/draft',\n",
       "    'description': 'Should be ignored by regular mortals.\\n\\nATM contains only a few files from http://github.com/dandi-datasets/nwb_test_data which more or less appropriate (do not lack critical metadata) for testing'},\n",
       "   'id': '000027'},\n",
       "  {'dc': {'title': 'Simulated cortical Neuropixels recording with ground truth',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000028/draft',\n",
       "    'description': 'A 10 minute Neuropixels recording simulated using the MEArec Python package (Buccino et al., 2020). The recording contains the spiking activity of 250 biophysically detailed neurons (200 excitatory and 50 inhibitory cells from the Neocortical Micro Circuit Portal with independent Poisson firing patterns, and additive Gaussian noise with 10uV standard deviation.'},\n",
       "   'id': '000028'},\n",
       "  {'dc': {'title': 'Test dataset for development purposes',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000029/draft',\n",
       "    'description': 'Should be ignored by regular mojgjhhj.  fjrtalddgdfgdfgs, and not relied upon being static or ever correct.\\nДолжно быть введено на Английском, но почему бы нам не проверить всю эту кухню \\nΔЙקم๗あ\\nabcddddd'},\n",
       "   'id': '000029'},\n",
       "  {'dc': {'title': 'Allen Brain Observatory Neuropixels recording',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000030/draft',\n",
       "    'description': 'A Neuropixels recording from a head-fixed mouse acquired at the Allen Institute for Brain Science. The recording has a duration of 15 minutes, 246 active recording channels (the remaining of the 384 Neuropixels channels were either not inserted in the brain tissue or below a firing rate of 0.1~Hz), and a sampling frequency of 30~kHz. The probe records from part of the cortex (V1), the hippocampus (CA1), the dentate gyrus, and the thalamus (LP).'},\n",
       "   'id': '000030'},\n",
       "  {'dc': {'title': 'ABC',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000031/draft',\n",
       "    'description': 'ABC'},\n",
       "   'id': '000031'},\n",
       "  {'dc': {'title': 'Test dataset',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000032/draft',\n",
       "    'description': 'Test description'},\n",
       "   'id': '000032'},\n",
       "  {'dc': {'title': 'Test-2 dataset',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000033/draft',\n",
       "    'description': 'Test-2 description'},\n",
       "   'id': '000033'},\n",
       "  {'dc': {'title': 'SpikeInterface, a unified framework for spike sorting',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000034/draft',\n",
       "    'description': 'SpikeInterface is a Python framework designed to make the analysis of extracellular electrophysiology data more accessible and standardized. The SpikeInterface documentation can be found at https://spikeinterface.readthedocs.io/en/latest/.\\n\\nThis dataset contains all recordings analysed in the paper: \"SpikeInterface, a unified framework for spike sorting.\" Alessio P. Buccino, Cole L. Hurwitz, Samuel Garcia, Jeremy Magland, Joshua H. Siegle, Roger Hurwitz, Matthias H. Hennig, eLife - doi: https://doi.org/10.7554/eLife.61834; '},\n",
       "   'id': '000034'},\n",
       "  {'dc': {'title': 'Temperature-controlled intracellular Patch-seq recordings in mouse motor cortex',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000035/draft',\n",
       "    'description': ' We used Patch-seq to combine patch-clamp recording, biocytin staining, and single-cell RNA sequencing of over 1300 neurons in adult mouse motor cortex, providing a comprehensive morpho-electric annotation of almost all transcriptomically defined neural cell types. This dataset contains rectangular stimulation of cells under temperature-controlled conditions (34 ℃). See Dandiset #8 for the main dataset, recorded under the room temperature.'},\n",
       "   'id': '000035'},\n",
       "  {'dc': {'title': 'Allen Institute Openscope - Measuring Stimulus-Evoked Neurophysiological Differentiation in Distinct Populations of Neurons in Mouse Visual Cortex',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000036/draft',\n",
       "    'description': 'Despite significant progress in understanding neural coding, it remains unclear how the coordinated activity of large populations of neurons relates to what an observer actually perceives. Since neurophysiological differences must underlie differences among percepts, differentiation analysis-quantifying distinct patterns of neurophysiological activity-has been proposed as an \"inside-out\" approach that addresses this question. This methodology contrasts with \"outside-in\" approaches such as feature tuning and decoding analyses, which are defined in terms of extrinsic experimental variables. Here, we used two-photon calcium imaging in mice of both sexes to systematically survey stimulus-evoked neurophysiological differentiation (ND) in excitatory neuronal populations in layers (L)2/3, L4, and L5 across five visual cortical areas (primary, lateromedial, anterolateral, posteromedial, and anteromedial) in response to naturalistic and phase-scrambled movie stimuli. We find that unscrambled stimuli evoke greater ND than scrambled stimuli specifically in L2/3 of the anterolateral and anteromedial areas, and that this effect is modulated by arousal state and locomotion. By contrast, decoding performance was far above chance and did not vary substantially across areas and layers. Differentiation also differed within the unscrambled stimulus set, suggesting that differentiation analysis may be used to probe the ethological relevance of individual stimuli.\\n\\nFor more details, consult the associated publication : https://doi.org/10.1523/eneuro.0280-21.2021'},\n",
       "   'id': '000036'},\n",
       "  {'dc': {'title': 'Allen Institute Openscope - Responses to inconsistent stimuli in somata and distal apical dendrites in primary visual cortex',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000037/draft',\n",
       "    'description': \"This dataset was collected for the Credit Assignment project, as part of the Allen Institute for Brain Science's OpenScope project. The most up-to-date links to the publications associated with this dataset (a dataset descriptor paper and an analysis paper) can be found by going to the metadata Github repository, linked below.\\n\\nBriefly, mice were habituated to head-fixation on a running disc over a period of 11 days. Over the last 6 days, mice were presented with repeating stimulus sequences, with consistent features. These sessions are not included in the dataset, as imaging was not performed. Following habituation, sequences with inconsistent features were introduced over three sessions, while two-photon calcium imaging was performed in pyramidal neurons of primary visual cortex (VisP). Specifically, recordings for each mouse were made in one of four planes: L2/3 somata, L5 somata, L2/3 distal apical dendrites or L5 distal apical dendrites. As mice gained experience with the inconsistent sequences, responses to these events were found to evolve in opposite ways in the dendritic and somatic compartments.\\n\\nThe dataset includes 50 sessions total, recorded in 13 subjects, with at least 3 sessions per subject. Each session file includes: **(1)** ROI dF/F traces, **(2)** ROI masks, **(3)** ROI tracking information (where applicable), **(4)** running velocity traces, **(5)** pupil diameter traces, **(6)** pupil centroid position traces in x and y, **(7)** stimulus parameters. Note that gaze was not computed for this dataset. What is provided is the position of the pupil centroid in the pupil recording videos. In addition, a second, slightly larger version of each file was created, also including: **(8)** stimulus frame images (identifiable by `+image` in the file name). Lastly, a third, much larger version of each file was created, also including: **(9)** the motion corrected imaging stack (identifiable by `_obj-raw` in the file name).\\n\\nCertain sessions in this dataset were excluded from our analyses for quality control reasons. They are provided here, as the reasons we excluded them from our analyses may not pose a problem for all use cases. The metadata Github repository, linked below, provides detailed information about each session. \\n\\nRelease notes for each version of the dataset are provided in the CHANGES file.\"},\n",
       "   'id': '000037'},\n",
       "  {'dc': {'title': 'Testytest',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000038/draft',\n",
       "    'description': 'I am just wondering what happens after I do thiis'},\n",
       "   'id': '000038'},\n",
       "  {'dc': {'title': 'Allen Institute – Contrast tuning in mouse visual cortex with calcium imaging',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000039/draft',\n",
       "    'description': 'A two photon calcium imaging dataset from Allen Institute measuring responses to full-field drifting gratings (approx. 120x90 degrees of visual space) of 8 directions and 6 contrasts (5%, 10%, 20%, 40%, 60%, 80%). Mouse Cre lines expressing GCaMP6f were imaged to record responses of   pyramidal neurons across cortical layers (Cux2: layer 2/3; Rorb: layer 4; Rbp4: layer 5; Ntsr1: layer 6) as well as inhibitory interneurons (Vip and Sst). All experimental sessions took place on the same data collection pipeline as the Allen Brain Observatory (see http://observatory.brain-map.org/visualcoding) and have the same visual stimulus monitor calibration and positioning, two photon imaging systems and image processing pipeline, and running wheel to track locomotion.\\n\\n Data are subject to Allen Institute Terms of Use policy, available at: http://www.alleninstitute.org/legal/terms-use/'},\n",
       "   'id': '000039'},\n",
       "  {'dc': {'title': 'Neuropixels recordings in mouse visual system',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000040/draft',\n",
       "    'description': 'This dataset contains electrophysiological recordings in mouse visual system collected at Allen Institute. '},\n",
       "   'id': '000040'},\n",
       "  {'dc': {'title': 'Network Homeostasis and State Dynamics of Neocortical Sleep',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000041/draft',\n",
       "    'description': 'Data was recorded using silicon probe electrodes in the frontal cortices of male Long Evans rats between 4-7 months of age. The design was to have no specific behavior, task or stimulus, rather the animal was left alone in it’s home cage (which it lives in at all\\ntimes). Data includes both local field potentials (LFP) and spikes. 11 total animals, 27 recording sessions, 1360 total units recorded, 1121 units considered stable, 995 putative excitatory units and 126 putative inhibitory units. Only recordings including a “WAKE-SLEEP” episode wherein at least 7 minutes of wake are followed by 20 minutes of sleep. On average 2 such WAKE-SLEEP episodes per recording session. '},\n",
       "   'id': '000041'},\n",
       "  {'dc': {'title': 'Allen Institute Openscope - Meaningful project  - Full movies',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000042/draft',\n",
       "    'description': ' The meaningful project characterized the preferences of mice visual cortical areas for natural images. This dataset contains full motion-corrected movies and each NWB file is therefore about 60GB.'},\n",
       "   'id': '000042'},\n",
       "  {'dc': {'title': 'Human, macaque, and mouse L5 pyramidal neuron physiology',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000043/draft',\n",
       "    'description': 'Patch clamp recordings from human premotor cortex, macaque M1 and mouse MOp.'},\n",
       "   'id': '000043'},\n",
       "  {'dc': {'title': 'Diversity in neural firing dynamics supports both rigid and learned hippocampal sequences',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000044/draft',\n",
       "    'description': 'This data set is composed of eight bilateral silicon-probe multi-cellular electrophysiological recordings performed on four male Long-Evans rats in the Buzsáki lab at NYU. These recordings were performed to assess the effect of novel spatial learning on hippocampal CA1 neural firing and LFP patterns in naïve animals. Each session consisted of a long (~4 hour) PRE rest/sleep epoch home-cage recordings performed in a familiar room, followed by a Novel MAZE running epoch (~45 minutes) in which the animals were transferred to a novel room, and water-rewarded to run on a novel maze. These mazes were either A) a wooden 1.6m linear platform, B) a wooden 1m diameter circular platform or C) a 2m metal linear platform. Animals were rewarded either at both ends of the linear platform, or at a predetermined location on the circular platform. The animal was gently encouraged to run unidirectionally on the circular platform. After the MAZE epochs the animals were transferred back to their home-cage in the familiar room where a long (~4 hour) POST rest/sleep was recorded. All eight sessions were concatenated from the PRE, MAZE, and POST recording epochs. In addition to hippocampal electrophysiological recordings, neck EMG and head-mounted accelerometer signals were recorded, and the animal’s position during MAZE running epochs was tracked via head-mounted LEDs.'},\n",
       "   'id': '000044'},\n",
       "  {'dc': {'title': 'IBL behavioral data',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000045/draft',\n",
       "    'description': 'This dataset is a result of a multi-institution, cross country collaboration of labs, called International Brain Laboratory conducting standardized experiments on decision-making in mice. In the experiment, mice are shown a visual grating on screen with varying levels of contrast, and they are trained to rotate a wheel to move the on-screen stimulus from either side of their visual field to the center. These NWB files contain datasets corresponding to this wheel movement, camera footage of mice and information about the visual stimulus at every trial. \\nPaper: \\nAguillon, V., Angelaki, D., Bayer, H. M., Bonacchi, N., Carandini, M., Cazettes, F., Churchland, A. K., Chapuis, G., Dan, Y., Dewitt, E., Faulkner, M., Hamish, F., Haetzel, L., Hausser, M., Hofer, S., Hu, F., Khanal, A., Krasniak, C., Laranjeira, I., … Zador, A. (2020). A standardized and reproducible method to measure decision-making in mice. BioRxiv, 2020.01.17.909838. https://doi.org/10.1101/2020.01.17.909838'},\n",
       "   'id': '000045'},\n",
       "  {'dc': {'title': 'asdf',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000046/draft',\n",
       "    'description': 'asdf'},\n",
       "   'id': '000046'},\n",
       "  {'dc': {'title': 'Test',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000047/draft',\n",
       "    'description': '...'},\n",
       "   'id': '000047'},\n",
       "  {'dc': {'title': 'Electrical and optical physiology in in vivo population-scale two-photon calcium imaging',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000048/draft',\n",
       "    'description': 'Spiking activity and simultaneously recorded fluorescence signals in transgenic mice expressing GCaMP6\\n\\nWe present a dataset consisting of simultaneously measured fluorescence and spiking activity of pyramidal neurons in layer 2/3 of primary visual cortex in transgenic mouse lines expressing genetically-encoded calcium indicators (GECIs) GCaMP6s or GCaMP6f.\\n\\nReference: https://portal.brain-map.org/explore/circuits/oephys'},\n",
       "   'id': '000048'},\n",
       "  {'dc': {'title': 'Allen Institute – TF x SF tuning in mouse visual cortex with calcium imaging',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000049/draft',\n",
       "    'description': 'A two photon calcium imaging dataset from Allen Institute measuring responses to full-field drifting gratings (approx. 120x90 degrees of visual space) moving in 4 different directions, at 5 spatial frequencies (0.02, 0.04, 0.08, 0.16, 0.32 cpd), and 5 temporal frequencies (0.5, 1, 2, 4, 8 Hz). The ratio of TF/SF is speed (deg/sec) and the extent to which visual neurons exhibit speed tuning has been shown to vary across some cortical areas (Andermann et al. 2011).\\nMouse Cre lines expressing GCaMPf were imaged to record responses of pyramidal neurons across cortical layers (Cux2: layer 2/3; Rorb: layer 4; Rbp4: layer 5; Ntsr1: layer 6) as well as somatostatin inhibitory interneurons (Sst). All Cre lines were imaged in VISp, and some (Cux2 and Sst) were also imaged in VISl, VISal, VISpm, VISam, and VISrl. All experimental sessions took place on the same data collection pipeline as the Allen Brain Observatory (see de Vries, Lecoq, Buice et al. 2020) and have the same visual stimulus monitor calibration and positioning, two photon imaging systems and image processing pipeline, and running wheel to track locomotion. Data are subject to Allen Institute Terms of Use policy, available at: http://www.alleninstitute.org/legal/terms-use/'},\n",
       "   'id': '000049'},\n",
       "  {'dc': {'title': 'Allen Institute - Run Tuning in the Mouse Visual Cortex',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000050/draft',\n",
       "    'description': 'Allen Institute for Brain Science, MindScope Project.'},\n",
       "   'id': '000050'},\n",
       "  {'dc': {'title': 'pons8-yo_16xdownsampled',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000051/draft',\n",
       "    'description': 'Downsampled image of pons slice #8 stained with a nuclear dye. For testing only.\\n'},\n",
       "   'id': '000051'},\n",
       "  {'dc': {'title': 'Pons8-BIDS-16xdownsampled',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000052/draft',\n",
       "    'description': 'A test upload of Pons8 YO and Lectin, downsampled 16x using the BIDS schema'},\n",
       "   'id': '000052'},\n",
       "  {'dc': {'title': 'Recordings from medial entorhinal cortex during linear track and open exploration',\n",
       "    'identifier': 'https://dandiarchive.org/dandiset/000053/draft',\n",
       "    'description': 'This dataset contains two task types. The first is tetrode recordings from medial entorhinal cortex during open field navigation with simultaneous inertial measurements of the head, and the second is Neuropixel recordings from medial entorhinal cortex during navigation down a virtual linear track with simultaneous eye measurements.'},\n",
       "   'id': '000053'}]}"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "datasets = k.get_datasets(datasource_list[0])\n",
    "datasets\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "4ae2a7be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'A challenge for data sharing in systems neuroscience is the multitude of different data formats used. Neurodata Without Borders: Neurophysiology 2.0 (NWB:N) has emerged as a standardized data format for the storage of cellular-level data together with meta-data, stimulus information, and behavior. A key next step to facilitate NWB:N adoption is to provide easy to use processing pipelines to import/export data from/to NWB:N. Here, we present a NWB-formatted dataset of 1863 single neurons recorded from the medial temporal lobes of 59 human subjects undergoing intracranial monitoring while they performed a recognition memory task. We provide code to analyze and export/import stimuli, behavior, and electrophysiological recordings to/from NWB in both MATLAB and Python. The data files are NWB:N compliant, which affords interoperability between programming languages and operating systems. This combined data and code release is a case study for how to utilize NWB:N for human single-neuron recordings and enables easy re-use of this hard-to-obtain data for both teaching and research on the mechanisms of human memory.'"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "z = datasets.get(\"results\")[1].get('dc').get(\"description\")\n",
    "z"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "b1daf335",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_details = k.get_dataset_details(\"scr_003510_cil_images\",\"gnCLTHgBhYBE05M-TuRO\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "8a6d0f37",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'processinghistory': 'film, micrograph, Print from negative scanned for Photoshop., recorded image',\n",
       " 'processinghistory_id': 'FBbi:00000265, FBbi:00000303',\n",
       " 'speciestaxaspecific': 'tetrahymena',\n",
       " 'v_uuid': '0420321d-0fc4-5e91-8489-39a99a0bcaf8',\n",
       " 'relationtointactcell_id': 'FBbi:00000029',\n",
       " 'pathologicalprocess': '',\n",
       " 'parameterimaged_id': 'FBbi:00000315',\n",
       " 'v_lastmodified_epoch': '1.459854817043704E9',\n",
       " 'id': '34735',\n",
       " 'dc': {'identifier': 'http://cellimagelibrary.org/images/34735',\n",
       "  'description': 'A micrograph of Tetrahymena showing a transverse ribbon starting from the anterior right side of the proximal end of a basal body (out of section plane) and crossing under the pellicle to end under the longitudinal ribbon  in the adjacent ridge crest. The epiplasm separates the two ribbons. Mitochondria often lie in ridges and the crescent end of kinetodesmal fiber, a dictyosome, and the endosomal membrane is seen. TEM taken on 8/25/67 by R. Allen with Philips 200 operating at 60kV. Neg. 19,200X. Bar = 0.2µm. The negative was printed to paper and the image was scanned to Photoshop. This digitized image is available for qualitative analysis. An unprocessed, high resolution version of this image (CIL:34728) is in the library and available for quantitative analysis. Additional information available at (http://www5.pbrc.hawaii.edu/allen/).',\n",
       "  'title': 'CIL:34735 - Tetrahymena pyriformis - cell by organism'},\n",
       " 'image_url': 'http://grackle.crbs.ucsd.edu:8001/ascb_il/render_thumbnail/34735/220/',\n",
       " 'v_uid': '34735',\n",
       " 'itemtype_id': 'FBbi:00000258, FBbi:00000273',\n",
       " 'pathologicalprocess_id': '',\n",
       " 'cellline': '',\n",
       " 'e_uid': '34735',\n",
       " 'v_lastmodified': '2017-01-18 16:59:51.670749',\n",
       " 'itemtype': 'illumination by electrons, transmission electron microscopy (TEM)',\n",
       " 'technicaldetails': 'Standard glutaraldehyde fixation followed by osmium tetroxide, dehydrated in alcohol and embedded in an epoxy resin. Microtome sections prepared at approximately 75nm thickness. Additional information available at (http://www5.pbrc.hawaii.edu/allen/).',\n",
       " 'termsandconditions': 'public_domain',\n",
       " 'cellline_id': '',\n",
       " 'relationtointactcell': 'microtome-sectioned tissue',\n",
       " 'parameterimaged': 'electron density',\n",
       " 'v_status': 'E',\n",
       " 'species': ['Tetrahymena pyriformis'],\n",
       " 'release_date': '2011-05-25',\n",
       " 'ncbiorganismalclassification_id': 'NCBITaxon:5908',\n",
       " 'biological_process': ['microtubule cytoskeleton organisation',\n",
       "  ' microtubule cytoskeleton organization'],\n",
       " 'cell_type': ['cell by organism',\n",
       "  ' Ciliated Protist',\n",
       "  ' eukaryotic cell',\n",
       "  ' Eukaryotic Protist'],\n",
       " 'visualization_method': ['gold',\n",
       "  ' positive staining',\n",
       "  ' stain with broad specificity',\n",
       "  ' uranyl salt'],\n",
       " 'data_qualification': ['PROCESSED'],\n",
       " 'biological_process_ids': ['GO:0000226'],\n",
       " 'preparation_ids': ['FBbi:00000011', ' FBbi:00000029', ' FBbi:00000089'],\n",
       " 'cellular_component': ['cell cortex',\n",
       "  ' cell periphery',\n",
       "  ' cortical cytoskeleton'],\n",
       " 'visualization_method_ids': ['FBbi:00000202',\n",
       "  ' FBbi:00000415',\n",
       "  ' FBbi:00000569'],\n",
       " 'attributions': ['Richard Allen'],\n",
       " 'celltype_ids': ['CL:0000004', ' CL:0000255'],\n",
       " 'cellular_component_id': ['CL:0000004', ' CL:0000255'],\n",
       " 'preparation_name': ['glutaraldehyde fixed tissue',\n",
       "  ' microtome-sectioned tissue',\n",
       "  ' monoclonal antibody',\n",
       "  ' tissue in amorphous ice embedment',\n",
       "  ' tissue in vitreous ice embedment'],\n",
       " 'imaging_mode': ['detection of electrons', ' film'],\n",
       " 'imaging_mode_id': ['FBbi:00000303', ' FBbi:00000375'],\n",
       " 'dimension_units': 'nanometers'}"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset_details.get(\"meta\").get(\"dc\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "1d893023",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'identifier': 'http://cellimagelibrary.org/images/11972',\n",
       " 'description': 'Preparation of the live ciliate Tetrahymena pyriformis stained with a mitochondria-specific probe (Mitotracker green) and viewed with the CSU-10 confocal microscope.  The mitochondria become visible as they swim through the confocal plane of focus.  Non-confocal background light was superimposed by backlighting through the condenser to make the whole organism visible.',\n",
       " 'title': 'CIL:11972 - Tetrahymena pyriformis - '}"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k.global_search_datasets('Tetrahymena pyriformis').get(\"results\")[0].get('dc')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "355d84de",
   "metadata": {},
   "outputs": [
    {
     "ename": "HTTPError",
     "evalue": "404 Client Error: Not Found for url: https://api.knowledge-space.org/cdes/age",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mHTTPError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[140]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m kg = \u001b[43mk\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget_cde\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mage\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[32m      2\u001b[39m kg\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\Swastik R\\Documents\\GitHub\\knowledge-space-agent\\backend\\knowledgespace_api.py:70\u001b[39m, in \u001b[36mget_cde\u001b[39m\u001b[34m(cde_id)\u001b[39m\n\u001b[32m     68\u001b[39m url = \u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mKS_BASE\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m/cdes/\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mcde_id\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m\"\u001b[39m\n\u001b[32m     69\u001b[39m response = requests.get(url, timeout=\u001b[32m10\u001b[39m)\n\u001b[32m---> \u001b[39m\u001b[32m70\u001b[39m \u001b[43mresponse\u001b[49m\u001b[43m.\u001b[49m\u001b[43mraise_for_status\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m     71\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m response.json()\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\Swastik R\\Documents\\GitHub\\knowledge-space-agent\\.venv\\Lib\\site-packages\\requests\\models.py:1026\u001b[39m, in \u001b[36mResponse.raise_for_status\u001b[39m\u001b[34m(self)\u001b[39m\n\u001b[32m   1021\u001b[39m     http_error_msg = (\n\u001b[32m   1022\u001b[39m         \u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m.status_code\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m Server Error: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mreason\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m for url: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m.url\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m\"\u001b[39m\n\u001b[32m   1023\u001b[39m     )\n\u001b[32m   1025\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m http_error_msg:\n\u001b[32m-> \u001b[39m\u001b[32m1026\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m HTTPError(http_error_msg, response=\u001b[38;5;28mself\u001b[39m)\n",
      "\u001b[31mHTTPError\u001b[39m: 404 Client Error: Not Found for url: https://api.knowledge-space.org/cdes/age"
     ]
    }
   ],
   "source": [
    "kg = k.get_cde(\"age\")\n",
    "kg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "f3a5c52d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'id': '7msPjHcBhYBE05M-qGUX',\n",
       "  'cde_ids': {'ninds': 'C00008', 'BRICS': 'AgeVal', 'nlm': 'jZT1bnHTnC8'},\n",
       "  'names': ['Age value',\n",
       "   \"(or age if you don't know date of birth)\",\n",
       "   'Age',\n",
       "   'Age (months)',\n",
       "   'Age at MRI',\n",
       "   'Age at dx',\n",
       "   'Age of affected',\n",
       "   'Age/DOB:',\n",
       "   'Age:',\n",
       "   'Participant Age when EEG obtained',\n",
       "   \"Pupil's Age\",\n",
       "   'Subject age:',\n",
       "   \"Subject's age (recorded in years)\",\n",
       "   \"Subject's age (recorded in years):\",\n",
       "   'Test Age'],\n",
       "  'definition': \"Value for participant/subject's age, calculated as elapsed time since the birth of the participant/subject.\",\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=jZT1bnHTnC8'},\n",
       " {'id': 'lGsPjHcBhYBE05M-5ZR-',\n",
       "  'cde_ids': {'ninds': 'C00002',\n",
       "   'BRICS': 'PostConceptnAgeVal',\n",
       "   'nlm': 'eYotanGYWFT'},\n",
       "  'names': ['Post conceptional age value',\n",
       "   'Post Conceptional Age (PCA)',\n",
       "   'Post conceptional age value (PCA)'],\n",
       "  'definition': 'Gestational age plus postnatal age of the participant/subject',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=eYotanGYWFT'},\n",
       " {'id': 'rWsPjHcBhYBE05M-wXcM',\n",
       "  'cde_ids': {'ninds': 'C19671',\n",
       "   'BRICS': 'FmlyHistCrrntDecesAgeVal',\n",
       "   'nlm': 'QJESwX5jSA'},\n",
       "  'names': ['Family history current decease age value',\n",
       "   'Current Age, or Age at Death'],\n",
       "  'definition': 'Value of current age or age at death for relative with the disease in question',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=QJESwX5jSA'},\n",
       " {'id': 'A2sPjHcBhYBE05M-u3NC',\n",
       "  'cde_ids': {'ninds': 'C10501',\n",
       "   'BRICS': 'DiagnosAgeVal',\n",
       "   'nlm': 'achaOkePd6G'},\n",
       "  'names': ['Diagnosis initial age value',\n",
       "   'Age at diagnosis',\n",
       "   'Age of Diagnosis'],\n",
       "  'definition': 'Age of the participant/subject when initially diagnosed with disease/disorder',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=achaOkePd6G'},\n",
       " {'id': '92sPjHcBhYBE05M-1IZj',\n",
       "  'cde_ids': {'ninds': 'C52710',\n",
       "   'BRICS': 'MenarcheBeginAgeVal',\n",
       "   'nlm': 'QypT1d9057'},\n",
       "  'names': ['Menarche begin age value', 'Age:'],\n",
       "  'definition': 'The age at which menarche began for the participant/subject',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=QypT1d9057'},\n",
       " {'id': 'n2sPjHcBhYBE05M-wXcM',\n",
       "  'cde_ids': {'ninds': 'C18205',\n",
       "   'BRICS': 'FmlyHistDMDBMDCrrntDecesAgeVal',\n",
       "   'nlm': 'q0wAq7gE2XJ'},\n",
       "  'names': ['Family history DMD BMD current decease age value',\n",
       "   'Current Age or Age at Death'],\n",
       "  'definition': 'Value of current age or age at death for relative with DMD or BMD',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=q0wAq7gE2XJ'},\n",
       " {'id': 'sGsPjHcBhYBE05M-wXcM',\n",
       "  'cde_ids': {'ninds': 'C19670',\n",
       "   'BRICS': 'FmlyHistDiagnsisAgeInd',\n",
       "   'nlm': 'QJHNeUVcsHC'},\n",
       "  'names': ['Family history diagnosis age indicator',\n",
       "   'Age of onset:',\n",
       "   'Age when diagnosed?'],\n",
       "  'definition': 'Indicator for whether the participant/subject knows the age of the relative when diagnosed',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=QJHNeUVcsHC'},\n",
       " {'id': 'QWsPjHcBhYBE05M-uXLx',\n",
       "  'cde_ids': {'ninds': 'C54722', 'BRICS': 'DeathAgeVal', 'nlm': 'XkM1X_5AqX'},\n",
       "  'names': ['Death age value', 'Age at death (years)'],\n",
       "  'definition': \"Value of the subject/participant's age at death\",\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=XkM1X_5AqX'},\n",
       " {'id': '62sPjHcBhYBE05M-qGUX',\n",
       "  'cde_ids': {'ninds': 'C19888',\n",
       "   'BRICS': 'AgeBiopsyYearVal',\n",
       "   'nlm': 'm1uUNmqoBC'},\n",
       "  'names': ['Age biopsy year value', 'Age at biopsy'],\n",
       "  'definition': 'Age (in years) of participant/subject at biopsy',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=m1uUNmqoBC'},\n",
       " {'id': '7WsPjHcBhYBE05M-qGUX',\n",
       "  'cde_ids': {'ninds': 'C17900',\n",
       "   'BRICS': 'AgePresentYearVal',\n",
       "   'nlm': 'FS1H85kEyHg'},\n",
       "  'names': ['Age presentation year value', 'Age at presentation'],\n",
       "  'definition': 'Age (in years) of participant/subject at presentation',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=FS1H85kEyHg'},\n",
       " {'id': 'y2sPjHcBhYBE05M-wXd3',\n",
       "  'cde_ids': {'ninds': 'C14944',\n",
       "   'BRICS': 'FmlyHistRelAgeVal',\n",
       "   'nlm': 'orgQMf7s3wo'},\n",
       "  'names': ['Family history relative age value', 'Current Age'],\n",
       "  'definition': \"Value for participant's/subject's family member's current age.\",\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=orgQMf7s3wo'},\n",
       " {'id': 'OmsPjHcBhYBE05M-24yT',\n",
       "  'cde_ids': {'ninds': 'C56304',\n",
       "   'BRICS': 'NIHTBCogPicVocAge3AdjSclScore',\n",
       "   'nlm': 'XJZ2RDd6aI'},\n",
       "  'names': ['NIH Toolbox Cognitive ? Picture vocabulary age 3 up age adjusted scale score',\n",
       "   'NIHTB Picture Vocabulary Age 3+ Age Adjusted Scale Score'],\n",
       "  'definition': 'The age adjusted scale score for the NIHTB Picture Vocabulary Age 3+ field of the NIH Toolbox Cognitive Battery.',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=XJZ2RDd6aI'},\n",
       " {'id': 'ZmsPjHcBhYBE05M-qGan',\n",
       "  'cde_ids': {'ninds': 'C00702',\n",
       "   'BRICS': 'AlcUseStrtAgeVal',\n",
       "   'nlm': 'FtNNFMKHCM4'},\n",
       "  'names': ['Alcohol use started age value',\n",
       "   'Age started drinking',\n",
       "   'Age started drinking (years):'],\n",
       "  'definition': 'Age in years when participant/subject started ingesting alcoholic beverages, including social drinking',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=FtNNFMKHCM4'},\n",
       " {'id': 'Z2sPjHcBhYBE05M-qGan',\n",
       "  'cde_ids': {'ninds': 'C00700',\n",
       "   'BRICS': 'AlcUseStopAgeVal',\n",
       "   'nlm': '2Cwxj7Q6xSP'},\n",
       "  'names': ['Alcohol use stopped age value',\n",
       "   'Age quit drinking',\n",
       "   'Age quit drinking (years):'],\n",
       "  'definition': 'Age in years when participant/subject stopped ingesting alcoholic beverages, including social drinking',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=2Cwxj7Q6xSP'},\n",
       " {'id': 'zWsPjHcBhYBE05M-96It',\n",
       "  'cde_ids': {'ninds': 'C10500',\n",
       "   'BRICS': 'SymptomOnsetAgeVal',\n",
       "   'nlm': 'bdRBYQ63dZ5'},\n",
       "  'names': ['Symptom onset age value',\n",
       "   'Age at Symptom Onset',\n",
       "   'If yes, indicate age when first noticed'],\n",
       "  'definition': 'Age when symptoms first experienced',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=bdRBYQ63dZ5'},\n",
       " {'id': 'yWsPjHcBhYBE05M-wXd3',\n",
       "  'cde_ids': {'ninds': 'C16146',\n",
       "   'BRICS': 'FmlyHistReltvAgeDeathVal',\n",
       "   'nlm': 'emxAyk2oF8y'},\n",
       "  'names': ['Family history relative age at death value',\n",
       "   'Age at Death',\n",
       "   'Age at death'],\n",
       "  'definition': \"Age at death of the participant/subject's relative, if applicable\",\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=emxAyk2oF8y'},\n",
       " {'id': '_WsPjHcBhYBE05M-1IZj',\n",
       "  'cde_ids': {'ninds': 'C52709',\n",
       "   'BRICS': 'MenopauseBeginAgeVal',\n",
       "   'nlm': 'QkKlFkdc0qm'},\n",
       "  'names': ['Menopause begin age value',\n",
       "   'Age:',\n",
       "   'If yes, age at beginning of menopause (years)'],\n",
       "  'definition': 'The age at which menopause began for the participant/subject',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=QkKlFkdc0qm'},\n",
       " {'id': 'E2sPjHcBhYBE05M-qWf6',\n",
       "  'cde_ids': {'BRICS': 'AnimalAgeGroup', 'nlm': 'QkguqizjN'},\n",
       "  'names': ['Animal age group type', 'Animal age group'],\n",
       "  'definition': [\"Type of age group the animal subject belongs to, at the time of injury. Value for animal participant's subject age, calculated as elapsed time since the birth of the participant/subject.\",\n",
       "   'Age group that the animal subject belongs to, at the time of injury.'],\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=QkguqizjN'},\n",
       " {'id': 'ymsPjHcBhYBE05M-wXd3',\n",
       "  'cde_ids': {'ninds': 'C16159',\n",
       "   'BRICS': 'FmlyHistRltveAgeOnsetVal',\n",
       "   'nlm': '1rQtjEA93n9'},\n",
       "  'names': ['Family history relative age at onset value',\n",
       "   'Age at first symptoms',\n",
       "   'Age at onset of MS',\n",
       "   'Age of onset'],\n",
       "  'definition': \"Age the participant's/subject's relative experienced the onset of the disease or disorder\",\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=1rQtjEA93n9'},\n",
       " {'id': '62sPjHcBhYBE05M-9aG4',\n",
       "  'cde_ids': {'ninds': 'C51672',\n",
       "   'BRICS': 'SubjectAgeVal',\n",
       "   'nlm': 'Q17NPxQc09X'},\n",
       "  'names': ['Subject age value', 'Chronological Age (year)'],\n",
       "  'definition': \"Value for the subject's age, calculated as the elapsed time since his/her birth in years.\",\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=Q17NPxQc09X'},\n",
       " {'id': 'OGsPjHcBhYBE05M-v3ao',\n",
       "  'cde_ids': {'ninds': 'C14479',\n",
       "   'BRICS': 'EplpsySyndLessSpecAgeRelTyp',\n",
       "   'nlm': 'RxsVj80BbRZ'},\n",
       "  'names': ['Epilepsy syndrome less specific age relationship type',\n",
       "   'Less Specific Age Relationship'],\n",
       "  'definition': 'Common epilepsy syndromes that have a less specific age or no association with a specific age on onset.',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=RxsVj80BbRZ'},\n",
       " {'id': 'NWsPjHcBhYBE05M-24yT',\n",
       "  'cde_ids': {'ninds': 'C56344',\n",
       "   'BRICS': 'NiHTBCogPicSeqMeAge3AASclScore',\n",
       "   'nlm': 'QJeE1_u6TI'},\n",
       "  'names': ['NIH Toolbox Cognitive ? Picture sequence memory age 3 up age adjusted scale score',\n",
       "   'NIHTB Picture Seuqence Memory Age 3+ Age Adjusted Scale Score'],\n",
       "  'definition': 'The age adjusted scale score for the NIHTB Picture Sequence Memory Age 3+ field of the NIH Toolbox Cognitive Battery.',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=QJeE1_u6TI'},\n",
       " {'id': 'YmsPjHcBhYBE05M-2YoG',\n",
       "  'cde_ids': {'ninds': 'C19885',\n",
       "   'BRICS': 'MotorFuncHistAgeLostFuncVal',\n",
       "   'nlm': '7yG88mqsHA'},\n",
       "  'names': ['Motor function history age lost function value',\n",
       "   'Age lost (years)'],\n",
       "  'definition': 'Age lost number',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=7yG88mqsHA'},\n",
       " {'id': 'AWsPjHcBhYBE05M-24wQ',\n",
       "  'cde_ids': {'ninds': 'C56305',\n",
       "   'BRICS': 'NIHTBCogPicVocAge3NatPrcntVal',\n",
       "   'nlm': 'X1M3Av_T6L'},\n",
       "  'names': ['NIH Toolbox Cognitive - Picture vocabulary age 3 up national age adjusted percentage value',\n",
       "   'NIHTB Picture Vocabulary Age 3+ National Percentile (age adjusted)'],\n",
       "  'definition': 'The national percentile (age adjusted) that the subject is in, with respect to the NIHTB Picture Vocabulary Age 3+ field of the NIH Toolbox Cognitive Battery.',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=X1M3Av_T6L'},\n",
       " {'id': 'x2sPjHcBhYBE05M-p2Wb',\n",
       "  'cde_ids': {'ninds': 'C16150',\n",
       "   'BRICS': 'AdoptedAgeVal',\n",
       "   'nlm': 'y3dh5VZXd1k'},\n",
       "  'names': ['Adopted age value', 'Age at time of adoption'],\n",
       "  'definition': 'Age at time of adoption if participant/subject was adopted',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=y3dh5VZXd1k'},\n",
       " {'id': 'MmsPjHcBhYBE05M-v3ao',\n",
       "  'cde_ids': {'ninds': 'C14472',\n",
       "   'BRICS': 'EpilepsySyndromeAgeOnsetTyp',\n",
       "   'nlm': 'dMyhShVaYT1'},\n",
       "  'names': ['Epilepsy syndrome age of onset type',\n",
       "   'Syndromes by Age of Onset'],\n",
       "  'definition': 'Age period during which the epilepsy syndrome first appeared',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=dMyhShVaYT1'},\n",
       " {'id': 'sWsPjHcBhYBE05M-wXcM',\n",
       "  'cde_ids': {'ninds': 'C19669',\n",
       "   'BRICS': 'FmlyHistDiagnsisAgeVal',\n",
       "   'nlm': 'QkVugS79oSA'},\n",
       "  'names': ['Family history diagnosis age value', 'Age when diagnosed?'],\n",
       "  'definition': 'Age at which relative was diagnosed with the disease in question',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=QkVugS79oSA'},\n",
       " {'id': 'zWsPjHcBhYBE05M-wXd3',\n",
       "  'cde_ids': {'ninds': 'C14945',\n",
       "   'BRICS': 'FmlyHistRelDeathAgeVal',\n",
       "   'nlm': 'QCAvQOuEKuT'},\n",
       "  'names': ['Family history relative death age value', 'Current Age'],\n",
       "  'definition': \"Value for participant's/subject's family member's age at death, if deceased.\",\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=QCAvQOuEKuT'},\n",
       " {'id': 'pmsPjHcBhYBE05M-0YTD',\n",
       "  'cde_ids': {'ninds': 'C18386',\n",
       "   'BRICS': 'LessThreYrAgeVal',\n",
       "   'nlm': 'Ocx8Dq2zzkC'},\n",
       "  'names': ['Less three years age value',\n",
       "   'Age if less than 3 years of age (in months)',\n",
       "   'If less than 3 years of age'],\n",
       "  'definition': 'Number of months from zero to thirty-six months of age.',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=Ocx8Dq2zzkC'},\n",
       " {'id': 'EmsPjHcBhYBE05M-9qI7',\n",
       "  'cde_ids': {'ninds': 'C08018', 'BRICS': 'SurgAgeVal', 'nlm': 'dar68E9nZTm'},\n",
       "  'names': ['Surgery age value',\n",
       "   'Admission Date (MM/YYYY) or Age (approximate) Discharge Date (MM/YYYY) or Age (approximate)',\n",
       "   'Age (approximate)',\n",
       "   'Date(s) - list all (MM/YYYY) or Age (approximate)'],\n",
       "  'definition': 'Age of the participant/subject at the time of the selected surgical procedure',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=dar68E9nZTm'},\n",
       " {'id': 'iWsPjHcBhYBE05M-zoLO',\n",
       "  'cde_ids': {'ninds': 'C12612',\n",
       "   'BRICS': 'IntlSymptmAgeVal',\n",
       "   'nlm': 'T7VmNtrbE79'},\n",
       "  'names': ['Initial symptom age value', 'Age at symptom onset'],\n",
       "  'definition': 'The age at which the participant/subject first noted a neuromuscular symptom',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=T7VmNtrbE79'},\n",
       " {'id': 'kGsPjHcBhYBE05M-zoLO',\n",
       "  'cde_ids': {'ninds': 'C52401', 'BRICS': 'InjAgeVal', 'nlm': 'm1MxOWjTaL'},\n",
       "  'names': ['Injury age value', 'Age at time of injury'],\n",
       "  'definition': \"The value of the subject's age at the time he/she was injured\",\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=m1MxOWjTaL'},\n",
       " {'id': 'S2sPjHcBhYBE05M--qU9',\n",
       "  'cde_ids': {'ninds': 'C00703',\n",
       "   'BRICS': 'TobcoUseStrtAgeVal',\n",
       "   'nlm': 'dv9Z34EO5gA'},\n",
       "  'names': ['Tobacco use started age value',\n",
       "   'Age started tobacco use',\n",
       "   'Age started tobacco use (years):'],\n",
       "  'definition': 'Age in years when participant/subject started using tobacco products (e.g., cigarettes, cigars, chewing tobacco or pipe)',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=dv9Z34EO5gA'},\n",
       " {'id': 'TGsPjHcBhYBE05M--qU9',\n",
       "  'cde_ids': {'ninds': 'C00701',\n",
       "   'BRICS': 'TobcoUseStopAgeVal',\n",
       "   'nlm': 'gYPyA7h2WPt'},\n",
       "  'names': ['Tobacco use stopped age value',\n",
       "   'Age stopped tobacco use',\n",
       "   'Age stopped tobacco use (years):'],\n",
       "  'definition': 'Age in years when participant/subject stopped using tobacco products (e.g., cigarettes, cigars, chewing tobacco or pipe)',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=gYPyA7h2WPt'},\n",
       " {'id': 'JGsPjHcBhYBE05M-24yT',\n",
       "  'cde_ids': {'ninds': 'C56321',\n",
       "   'BRICS': 'NIHTBCogLisScWMeAge7AASclScore',\n",
       "   'nlm': 'm1b11_OTTI'},\n",
       "  'names': ['NIH Toolbox Cognitive ? List scoring working memory age 7 up age adjusted scale score',\n",
       "   'NIHTB List Sorting Word Memory Age 7+ Age Adjusted Scale Score'],\n",
       "  'definition': 'The age adjusted scale score for the NIHTB List Scoring Working Memory Age 7+ field of the NIH Toolbox Cognitive Battery.',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=m1b11_OTTI'},\n",
       " {'id': 'KWsPjHcBhYBE05M-24yT',\n",
       "  'cde_ids': {'ninds': 'C56350',\n",
       "   'BRICS': 'NIHTBCogORdReENGAge7AASclScore',\n",
       "   'nlm': 'QyMSyu_a6I'},\n",
       "  'names': ['NIH Toolbox Cognitive ? Oral reading recognition english age 7 up age adjusted scale score',\n",
       "   'NIHTB Oral Reading Recognition ENG Age 7+ Age Adjusted Scale Score'],\n",
       "  'definition': 'The age adjusted scale score for the NIHTB Oral Reading Recognition ENG Age 7+ field of the NIH Toolbox Cognitive Battery.',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=QyMSyu_a6I'},\n",
       " {'id': 'L2sPjHcBhYBE05M-24yT',\n",
       "  'cde_ids': {'ninds': 'C56336',\n",
       "   'BRICS': 'NIHTBCogPaCoPrSpAge7AASclScore',\n",
       "   'nlm': '7kzf1_uTTU'},\n",
       "  'names': ['NIH Toolbox Cognitive ? Pattern comparison processing speed age 7 up age adjusted scale score',\n",
       "   'NIHTB Pattern Comparison Processing Speed Age 7+ Age Adjusted Scale Score'],\n",
       "  'definition': 'The age adjusted scale score for the NIHTB Pattern Comparison Processing Speed Age 7+ field of the NIH Toolbox Cognitive Battery.',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=7kzf1_uTTU'},\n",
       " {'id': 'WmsPjHcBhYBE05M-xXvJ',\n",
       "  'cde_ids': {'ninds': 'C17411',\n",
       "   'BRICS': 'GestAgeWksDelivrDur',\n",
       "   'nlm': '7Xo4kdwo2GE'},\n",
       "  'names': ['Gestational age weeks delivery duration', 'Gestational age (GA)'],\n",
       "  'definition': \"Duration of gestation (gestational age or GA) at the time of the participant/subject's delivery\",\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=7Xo4kdwo2GE'},\n",
       " {'id': 'AGsPjHcBhYBE05M-24wQ',\n",
       "  'cde_ids': {'ninds': 'C56345',\n",
       "   'BRICS': 'NIHTBCogPicSMemAge3NatPrcntVal',\n",
       "   'nlm': 'myWVkOOpa8'},\n",
       "  'names': ['NIH Toolbox Cognitive - Picture sequence memory age 3 up national age adjusted percentage value',\n",
       "   'NIHTB Picture Seuqence Memory Age 3+ National Percentile (age adjusted)'],\n",
       "  'definition': 'The national percentile (age adjusted) that the subject is in, with respect to the NIHTB Picture Sequence memory Age 3+ field of the NIH Toolbox Cognitive Battery.',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=myWVkOOpa8'},\n",
       " {'id': 'wmsPjHcBhYBE05M-96It',\n",
       "  'cde_ids': {'ninds': 'C18243',\n",
       "   'BRICS': 'SymptmFrstApprAgTyp',\n",
       "   'nlm': '1QEtnd2wE_F'},\n",
       "  'names': ['Symptom first appear age type', 'Age at first symptom'],\n",
       "  'definition': 'Type of age of the participant/subject at first symptom appearance',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=1QEtnd2wE_F'},\n",
       " {'id': 'rWsPjHcBhYBE05M-xHrs',\n",
       "  'cde_ids': {'ninds': 'C12204',\n",
       "   'BRICS': 'GITherapyFundoplicationAgeVal',\n",
       "   'nlm': 'Hba8jTaC2A6'},\n",
       "  'names': ['Gastrointestinal therapy fundoplication age value',\n",
       "   '5. Fundoplication? If Yes, b. Age:',\n",
       "   'Fundoplication? If Yes, b. Age:'],\n",
       "  'definition': \"Value of the participant/subject's age at which s/he had fundoplication surgery\",\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=Hba8jTaC2A6'},\n",
       " {'id': 'H2sPjHcBhYBE05M-24yT',\n",
       "  'cde_ids': {'ninds': 'C56314',\n",
       "   'BRICS': 'NIHTBCogFIConAttAg3AdjSclScore',\n",
       "   'nlm': 'Xy0RPdTTI'},\n",
       "  'names': ['NIH Toolbox Cognitive ? Flanker inhibitory control and attention age 3 up age adjusted scale score',\n",
       "   'NIHTB Flanker Inhibitory Contorl and Attn Age 3+ Age Adjusted Score'],\n",
       "  'definition': 'The age adjusted scale score for the NIHTB Flanker Inhibitory Control and Attn Age 3+ field of the NIH Toolbox Cognitive Battery.',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=Xy0RPdTTI'},\n",
       " {'id': 'CGsPjHcBhYBE05M-yX5u',\n",
       "  'cde_ids': {'ninds': 'C57248',\n",
       "   'BRICS': 'HypothyDiagnosAgeVal',\n",
       "   'nlm': 'mJv_wtTTL'},\n",
       "  'names': ['Hypothyroidism diagnosis age value', 'If yes, age at diagnosis'],\n",
       "  'definition': 'The value representing age the subject was when he/she was diagnosed with hypothyroidism.',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=mJv_wtTTL'},\n",
       " {'id': '12sPjHcBhYBE05M-96It',\n",
       "  'cde_ids': {'ninds': 'C18242',\n",
       "   'BRICS': 'SymptmFrstApprAgVal',\n",
       "   'nlm': '7t2grb0nqwB'},\n",
       "  'names': ['Symptoms first appeared age value', 'Age at first symptom'],\n",
       "  'definition': 'Value of the age at which the participant/subject first noted a neuromuscular symptom',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=7t2grb0nqwB'},\n",
       " {'id': 'oWsPjHcBhYBE05M-wXcM',\n",
       "  'cde_ids': {'ninds': 'C18201',\n",
       "   'BRICS': 'FmlyHistDMDBMDDiagnsisAgeVal',\n",
       "   'nlm': 'nX3a4vDP36X'},\n",
       "  'names': ['Family history DMD BMD diagnosis age value',\n",
       "   'Age when diagnosed?'],\n",
       "  'definition': 'Age at which relative was diagnosed with DMD or BMD',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=nX3a4vDP36X'},\n",
       " {'id': 'iWsPjHcBhYBE05M-yH2R',\n",
       "  'cde_ids': {'ninds': 'C20018',\n",
       "   'BRICS': 'HormoneRltdSyndmDiagAgVal',\n",
       "   'nlm': 'QkAklS7qorC'},\n",
       "  'names': ['Hormone-related syndrome diagnosis age value',\n",
       "   'Age at diagnosis'],\n",
       "  'definition': 'The age value of participant/subject related to known hormone-related/endocrine syndromes or disorders',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=QkAklS7qorC'},\n",
       " {'id': '9GsPjHcBhYBE05M-yH30',\n",
       "  'cde_ids': {'ninds': 'C57246',\n",
       "   'BRICS': 'HyperthyDiagnosAgeVal',\n",
       "   'nlm': 'XJ_8WPtp6I'},\n",
       "  'names': ['Hyperthyroidism diagnosis age value', 'If yes, age at diagnosis'],\n",
       "  'definition': 'The value representing the age the subject was when he/she was diagnosed with hyperthyroidism.',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=XJ_8WPtp6I'},\n",
       " {'id': '_WsPjHcBhYBE05M-24sQ',\n",
       "  'cde_ids': {'ninds': 'C56329',\n",
       "   'BRICS': 'NIHTBCogDCCSAge3NatPrcntVal',\n",
       "   'nlm': 'my_bkdOa6I'},\n",
       "  'names': ['NIH Toolbox Cognitive - Dimensional change card sort age 3 up national age adjusted percentage value',\n",
       "   'NIHTB Dimensional Charge Card Sort Age 3+ National Percentile (age adjusted)'],\n",
       "  'definition': 'The national percentile (age adjusted) that the subject is in, with respect to the NIHTB Dimensional Change Card Sort Age 3+ field of the NIH Toolbox Cognitive Battery.',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=my_bkdOa6I'},\n",
       " {'id': '_2sPjHcBhYBE05M-24sQ',\n",
       "  'cde_ids': {'ninds': 'C56322',\n",
       "   'BRICS': 'NIHTBCogLSWMAge7NatPrcntVal',\n",
       "   'nlm': 'XkeyO_pT8'},\n",
       "  'names': ['NIH Toolbox Cognitive - List scoring working memory age 7 up national age adjusted percentage value',\n",
       "   'NIHTB List Sorting Word Memory Age 7+ National Percentile (age adjusted)'],\n",
       "  'definition': 'The national percentile (age adjusted) that the subject is in, with respect to the NIHTB List Scoring Working Memory Age 7+ field of the NIH Toolbox Cognitive Battery.',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=XkeyO_pT8'},\n",
       " {'id': 'LGsPjHcBhYBE05M-24yT',\n",
       "  'cde_ids': {'ninds': 'C56351',\n",
       "   'BRICS': 'NIHTBCogORREAge7NatPrcntVal',\n",
       "   'nlm': 'Q1IJdda6L'},\n",
       "  'names': ['NIH Toolbox Cognitive ? Oral reading recognition english age 7 up national age adjusted percentage value',\n",
       "   'NIHTB Oral Reading Recognition ENG Age 7+ National Percentile (age adjusted)'],\n",
       "  'definition': 'The national percentile (age adjusted) that the subject is in, with respect to the NIHTB Oral Reading Recognition ENG Age 7+ field of the NIH Toolbox Cognitive Battery.',\n",
       "  'url': 'https://cde.nlm.nih.gov/deView?tinyId=Q1IJdda6L'}]"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "KS_BASE = \"https://api.knowledge-space.org\"\n",
    "def get_cde(cde_id: str) -> Dict:\n",
    "    \"\"\"Get Common Data Element information\"\"\"\n",
    "    url = f\"{KS_BASE}/cde/search?q={cde_id}\"\n",
    "    response = requests.get(url, timeout=10)\n",
    "    response.raise_for_status()\n",
    "    return response.json()\n",
    "Kg = get_cde(\"age\")\n",
    "kg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "7bab3bc8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': '7msPjHcBhYBE05M-qGUX',\n",
       " 'cde_ids': {'ninds': 'C00008', 'BRICS': 'AgeVal', 'nlm': 'jZT1bnHTnC8'},\n",
       " 'names': ['Age value',\n",
       "  \"(or age if you don't know date of birth)\",\n",
       "  'Age',\n",
       "  'Age (months)',\n",
       "  'Age at MRI',\n",
       "  'Age at dx',\n",
       "  'Age of affected',\n",
       "  'Age/DOB:',\n",
       "  'Age:',\n",
       "  'Participant Age when EEG obtained',\n",
       "  \"Pupil's Age\",\n",
       "  'Subject age:',\n",
       "  \"Subject's age (recorded in years)\",\n",
       "  \"Subject's age (recorded in years):\",\n",
       "  'Test Age'],\n",
       " 'definition': \"Value for participant/subject's age, calculated as elapsed time since the birth of the participant/subject.\",\n",
       " 'url': 'https://cde.nlm.nih.gov/deView?tinyId=jZT1bnHTnC8'}"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kg[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "483e0936",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "k_Agent-main",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
